vck/src/commonMain/kotlin/at/asitplus/wallet/lib/data/dif/PresentationExchangeInputEvaluator.kt

```diff
@@ -0,0 +1,222 @@
+package at.asitplus.wallet.lib.data.dif
+
+import at.asitplus.KmmResult
+import at.asitplus.KmmResult.Companion.wrap
+import at.asitplus.dif.Constraint
+import at.asitplus.dif.ConstraintField
+import at.asitplus.dif.ConstraintFilter
+import at.asitplus.dif.FormatHolder
+import at.asitplus.dif.InputDescriptor
+import at.asitplus.dif.RequirementEnum
+import at.asitplus.jsonpath.JsonPath
+import at.asitplus.jsonpath.core.NodeList
+import at.asitplus.jsonpath.core.NormalizedJsonPath
+import at.asitplus.openid.CredentialFormatEnum
+import io.github.aakira.napier.Napier
+import kotlinx.serialization.json.JsonArray
+import kotlinx.serialization.json.JsonElement
+import kotlinx.serialization.json.JsonNull
+import kotlinx.serialization.json.JsonObject
+import kotlinx.serialization.json.JsonPrimitive
+import kotlinx.serialization.json.booleanOrNull
+import kotlinx.serialization.json.doubleOrNull
+import kotlinx.serialization.json.longOrNull
+
+/**
+ * Specification: https://identity.foundation/presentation-exchange/spec/v2.0.0/#input-evaluation
+ */
+object PresentationExchangeInputEvaluator {
+    fun evaluateInputDescriptorAgainstCredential(
+        inputDescriptor: InputDescriptor,
+        fallbackFormatHolder: FormatHolder?,
+        credentialClaimStructure: JsonElement,
+        credentialFormat: CredentialFormatEnum,
+        credentialScheme: String?,
+        pathAuthorizationValidator: (NormalizedJsonPath) -> Boolean,
+    ): KmmResult<Map<ConstraintField, NodeList>> = runCatching {
+        (inputDescriptor.format ?: fallbackFormatHolder)?.let { formatHolder ->
+            when(credentialFormat) {
+                CredentialFormatEnum.JWT_VC -> formatHolder.jwtVp != null
+                CredentialFormatEnum.VC_SD_JWT,
+                CredentialFormatEnum.DC_SD_JWT -> formatHolder.jwtSd != null || formatHolder.sdJwt != null
+                CredentialFormatEnum.MSO_MDOC -> formatHolder.msoMdoc != null
+                else -> false
+            }.let { isMatchingCredentialFormat ->
+                if(!isMatchingCredentialFormat) {
+                    throw InvalidCredentialFormatException()
```

Comment:
nitpick: Add message to exception to improve stacktraces? e.g. what was expected, and what was really there